### PR TITLE
Raster rendering rule

### PR DESCRIPTION
**Pull request by JonLavi**
Hi @tschie, could you please review? Inspired by the [android sample](https://github.com/Esri/arcgis-runtime-samples-android/tree/master/java/raster-rendering-rule).

### CODE REVIEW

File: src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java
Comment (tschie) [inline src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java]: This can be done in the start method. Android samples like to add unnecessary single use methods everywhere for some reason.

File: src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java
Comment (tschie) [inline src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java]: This also doesn't need to be a separate method.

File: src/main/java/com/esri/samples/raster/raster_rendering_rule/README.md
Comment (tschie) [inline src/main/java/com/esri/samples/raster/raster_rendering_rule/README.md]: Needs to be HTML

File: src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java
Comment (tschie) [inline src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java]: I noticed the rendering rule has a description as well which is pretty useful. Maybe that could be displayed in a label somewhere? Would help users know what the rendering rule is doing.

File: src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java
Comment (tschie) [inline src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java]: You could do `ComboBox<RenderingRuleInfo>`. Then set a converter on the combobox to show its name:
```java
comboBox.setConverter(new StringConverter...)`
```

Other samples do this. Then you don't have to keep around a reference to the list or use indices to get back the selected item.

File: src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java
Comment (tschie) [inline src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java]: I don't think this is needed. The map will auto-load the layer, and the layer will auto-load the raster.

File: src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java
Comment (tschie) [inline src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java]: Call this `renderingRuleComboBox` or `renderingRuleInfoComboBox`. Our convention is to have the UI control name as the last part of the variable name.

File: src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java
Comment (tschie) [inline src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java]: You don't need a variable for this. Just use the string directly as an argument.

**Review by tschie — CHANGES_REQUESTED** [review]
Looks good! Just a few convention things that need updating.

File: src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java
Comment (JonLavi) [reply src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java]: Good call! I've built a `VBox` for the `ComboBox` and `Label` for the description to live in. Would be nice to have the `VBox` resize dynamically to fit the text. Any tips on how that could be done?

**Comment by JonLavi** [comment]
Thanks, all done! Looking forward to seeing what you think.

File: src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java
Comment (Rachael-E) [inline src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java]: ArcGISMap should be a local variable: it doesn't need to be a member variable as it it not used outside of the start method.

File: src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java
Comment (Rachael-E) [inline src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java]: remove this line, not needed: by default the combobox is uneditable.

File: src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java
Comment (Rachael-E) [inline src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java]: With `new StringConverter<RenderingRuleInfo>`, this can be replaced with `new StringConverter<>`. IntelliJ should show RenderingRuleInfo between the <> as greyed out as a guide to show it's not required.

File: src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java
Comment (Rachael-E) [inline src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java:59]: if you add `      scene.getStylesheets().add(getClass().getResource("/css/style.css").toExternalForm());
` in here, then you can remove the `setFont` and `setTextFill` methods from your labels on line 102 and 103.

File: src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java
Comment (Rachael-E) [inline src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java]: setFont and setTextFill should be removed once the link to style.css is added

File: src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java
Comment (Rachael-E) [inline src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java]: Since you use this string twice in the code, I'd store it as a local variable. @tschie is this ok for convention?

File: src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java
Comment (Rachael-E) [inline src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java]: Can remove prompt text since the raster with a rule (albeit "none") is already loaded (see later comment to select "none" from dropdown box initially).

File: src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java
Comment (Rachael-E) [inline src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java:163]: can add `renderingRuleInfoComboBox.getSelectionModel().selectFirst();` in here to show the user the displayed raster on loading is with the rule "none".

File: src/main/java/com/esri/samples/raster/raster_rendering_rule/README.md
Comment (Rachael-E) [inline src/main/java/com/esri/samples/raster/raster_rendering_rule/README.md]: I'd add "raster" in here too.

File: src/main/java/com/esri/samples/raster/raster_rendering_rule/README.md
Comment (Rachael-E) [inline src/main/java/com/esri/samples/raster/raster_rendering_rule/README.md]: Relevant API sections should be in alphabetical order

File: src/main/java/com/esri/samples/raster/raster_rendering_rule/README.md
Comment (Rachael-E) [inline src/main/java/com/esri/samples/raster/raster_rendering_rule/README.md]: This part of the readme is intended as a quick summary of what the user wants to do. Framed as "I want to...":

Suggested edit: "Display a raster on a map and apply different rendering rules to that raster.

File: src/main/java/com/esri/samples/raster/raster_rendering_rule/README.md
Comment (Rachael-E) [inline src/main/java/com/esri/samples/raster/raster_rendering_rule/README.md:6]: Add Use case section here, as per your new readme design for this sample

File: src/main/java/com/esri/samples/raster/raster_rendering_rule/README.md
Comment (Rachael-E) [inline src/main/java/com/esri/samples/raster/raster_rendering_rule/README.md:16]: Ordered list, including methods specific to Java where appropriate. e.g. this is a good one to mention when describing getting the rendering rule info from the image service raster.
`imageServiceRaster.getServiceInfo().getRenderingRuleInfos();`

**Review by Rachael-E — CHANGES_REQUESTED** [review]
Great start to samples work! I've added a few more things following Tyler's review for you to have a look over.

File: src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java
Comment (JonLavi) [reply src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java]: It does indeed. Wasn't sure whether to opt for the verbose syntax here, but happy to keep it nice and tidy.

**Comment by JonLavi** [comment]
Thanks @Rachael-E. All suggestions integrated and readme updated. The sample now automatically selects the first rendering rule into the drop-down menu. There is still the slight delay with the loading of the raster, but a progress indicator now shows while loading.

File: src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java
Comment (tschie) [inline src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java]: This is fine. Sorry for the confusion with my earlier comment. This API seems to use a pattern I have no come across in our API before: You have to load an image service raster to see the options, and then you can't reuse that raster when choosing another option. Weird design IMO.

It should however be `ImageServiceRasterUri` used here (actually `ImageServiceUrl` would probably be more accurate). All caps should only be used for compile-time constants (**static** final variables). You could make this a static field if you want. That's what we've done in other samples.

File: src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java
Comment (tschie) [inline src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java]: Use the VBox constructor which takes a spacing argument instead.

File: src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java
Comment (tschie) [inline src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java]: Use `renderingRuleInfoComboBox.getItems().addAll(renderingRuleInfos);` instead

File: src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java
Comment (tschie) [inline src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java]: Should be `// clear all operational layers`. But that doesn't add any context. Maybe something like: `// clear raster layer using previous rendering rule`

File: src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java
Comment (tschie) [inline src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java]: Don't think this is needed. Its width should be constrained by the VBox it's inside.

File: src/main/java/com/esri/samples/raster/raster_rendering_rule/README.md
Comment (tschie) [inline src/main/java/com/esri/samples/raster/raster_rendering_rule/README.md]: > the interface for the selection and application of the required rule can be used

I don't know what this means. Please reach out to the team for a real-world use case.

File: src/main/java/com/esri/samples/raster/raster_rendering_rule/README.md
Comment (tschie) [inline src/main/java/com/esri/samples/raster/raster_rendering_rule/README.md:17]: This section is meant to be a high-level overview of the steps required to use the API. It should avoid specifics about how its being shown in the sample (anything about the UI). Here's what I'd suggest:

1. Create an `ImageServiceRaster` using a URL to an online image service.
2. After loading the raster, use `imageServiceRaster.getServiceInfo().getRenderingRuleInfos()` to get a list of `RenderingRuleInfo`s supported by the service.
3. Choose a rendering rule info to apply and use it to create a `RenderingRule`.
4. Create a new `ImageServiceRaster` using the same URL.
5. Apply the rendering rule to the new raster using `imageServiceRaster.setRenderingRule(renderingRuleInfo)`
6. Create a `RasterLayer` from the raster for display.

Since this sample does not have a common design readme, please open a PR with the markdown version of this so it can be reviewed by the team.

File: src/main/java/com/esri/samples/raster/raster_rendering_rule/README.md
Comment (tschie) [inline src/main/java/com/esri/samples/raster/raster_rendering_rule/README.md]: ~~covering North Carolina’s, City of Charlotte downtown area~~

covering Charlotte, North Carolina's downtown area

File: src/main/java/com/esri/samples/raster/raster_rendering_rule/README.md
Comment (tschie) [inline src/main/java/com/esri/samples/raster/raster_rendering_rule/README.md]: A `signal` is Qt-specific. In the Java API, this would be a callback.

This section is supposed to be higher-level than that though. You can simply write "load the image service raster."

**Review by tschie — APPROVED** [review]
This looks good. Only a few more minor comments. Please fix them up and then feel free to merge. You don't have to wait for approval on the common readme design PR, but please do make one (it will be good practice).

File: src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java
Comment (JonLavi) [reply src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java]: So much neater, great spot!

File: src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java
Comment (JonLavi) [reply src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java]: Opted for `// clear previous rendering rule from all operational layers`. What do you think?

File: src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java
Comment (JonLavi) [reply src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java]: It is not! I think I ended up having both because I tried `.setMaxWidth()` assuming it'll cause the text to wrap. Good to know.

File: src/main/java/com/esri/samples/raster/raster_rendering_rule/README.md
Comment (JonLavi) [reply src/main/java/com/esri/samples/raster/raster_rendering_rule/README.md:17]: Thanks, I've been working on a few readmes recently and slowly getting the feel for it. The more I work with the API the easier it gets to describe the steps. I'm seeing now the reference to the ComboBox is not ideal.

File: src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java
Comment (Rachael-E) [inline src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java]: I'd move this to line 109, just above here you first use it. The first part of the start method should always be concerned with the JavaFX set up (boilerplate) before getting into sample specific code.

File: src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java
Comment (Rachael-E) [inline src/main/java/com/esri/samples/raster/raster_rendering_rule/RasterRenderingRuleSample.java]: this line of code is clearing the operational layers from the map: i.e. it is removing the raster layer that you have placed within the operational layer each time. I'd use:

`// clear previous raster layer from the map's operational layers`

File: src/main/java/com/esri/samples/raster/raster_rendering_rule/README.md
Comment (Rachael-E) [reply src/main/java/com/esri/samples/raster/raster_rendering_rule/README.md]: I'd go for something like this. When you make a PR for this in the Common Samples repo, the team may have other suggestions!

"A single raster image can be rendered in a number of different ways, including representation of slope aspect, hillshade and shaded relief. Applying these different renderers to the same raster allows for a powerful visual analysis of the data. For example, a geologist could interrogate the raster image using all available renderers to map subtle geological features on a landscape not apparent in a regular aerial photograph."

Comment (JonLavi) [reply src/main/java/com/esri/samples/raster/raster_rendering_rule/README.md]: Very well put, thanks!

**Review by Rachael-E — APPROVED** [review]
This is all looking great @JonLavi! You can go ahead and merge now.

**Pull request merged**
